lib/lambda_ethereum_consensus/state_transition/operations.ex

```diff
@@ -0,0 +1,189 @@
+              %Withdrawal{index: index} = latest_withdrawal
+              %BeaconState{state | next_withdrawal_index: index + 1}
+            else
+              state
+            end
+
+          max_withdrawals_per_payload = ChainSpec.get("MAX_WITHDRAWALS_PER_PAYLOAD")
+
+          # Update the next validator index to start the next withdrawal sweep
+          response =
+            if length(expected_withdrawals) == max_withdrawals_per_payload do
+              # Next sweep starts after the latest withdrawal's validator index
+              latest_withdrawal = List.last(expected_withdrawals)
+              %Withdrawal{validator_index: validator_index} = latest_withdrawal
+              next_validator_index = rem(validator_index + 1, length(validators))
+
+              newer_state = %BeaconState{
+                new_state
+                | next_withdrawal_validator_index: next_validator_index
+              }
+
+              {:ok, newer_state}
+            else
+              # Advance sweep by the max length of the sweep if there was not a full set of withdrawals
+              max_validators_per_withdrawals_sweep =
+                ChainSpec.get("MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP")
+
+              next_index = next_withdrawal_validator_index + max_validators_per_withdrawals_sweep
+              next_validator_index = rem(next_index, length(validators))
+
+              newer_state = %BeaconState{
+                new_state
+                | next_withdrawal_validator_index: next_validator_index
+              }
+
+              {:ok, newer_state}
+            end
+
+          response
+
+        {:error, reason} ->
+          {:error, reason}
+      end
+
+    response
+  end
+
+  @spec decrease_balances(BeaconState.t(), list(Withdrawal.t()), list(Withdrawal.t())) ::
+          {:ok, BeaconState.t()} | {:error, String.t()}
+  defp decrease_balances(_state, withdrawals, expected_withdrawals)
+       when length(withdrawals) !== length(expected_withdrawals) do
+    {:error, "length of withdrawals is not equal to expected withdrawals"}
+  end
+
+  defp decrease_balances(
+         state,
+         withdrawals,
+         expected_withdrawals
+       ) do
+    result =
+      Enum.reduce_while(Enum.zip(expected_withdrawals, withdrawals), {:ok, state}, fn element,
+                                                                                      {_, state} ->
+        element_as_list = element |> Tuple.to_list()
+        expected_withdrawal = Enum.fetch!(element_as_list, 0)
+        withdrawal = Enum.fetch!(element_as_list, 1)
+
+        %Withdrawal{validator_index: validator_index, amount: amount} = withdrawal
+
+        if withdrawal !== expected_withdrawal do
+          {:halt, {:error, "withdrawal != expected_withdrawal"}}
+        else
+          new_state = Mutators.decrease_balance(state, validator_index, amount)
+          {:cont, {:ok, new_state}}
+        end
+      end)
+
+    result
```

Comment:
nit: let's refactor this a little bit:

```elixir
defp decrease_balances(
         state,
         withdrawals,
         expected_withdrawals
       ) do
  Enum.zip(expected_withdrawals, withdrawals)
  |> Enum.reduce_while({:ok, state}, &decrease_or_halt/2)
end
```

And then we can define a `decrease_or_halt` function:

```elixir
# If the withdrawals are the same, decrease the balance of the validator.
defp decrease_or_halt({withdrawal, withdrawal}, {:ok, state}) do
 {:cont, {:ok, Mutators.decrease_balance(state, withdrawal.validator_index, withdrawal.amount)}}
end

defp decrease_or_halt(_, _), do: {:halt, {:error, "withdrawal != expected_withdrawal"}}
```

The idea here is to:
- Remove the "results" intermediate variables
- Pipe instead of creating intermediate variables
- Use clauses instead of ifs, also using the clause to check equality between withdrawals.